src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingListObjectsOperation.cs

```diff
@@ -0,0 +1,219 @@
+                {
+                    ScriptingErrorParams eventParams = new ScriptingErrorParams
+                    {
+                        OperationId = this.OperationId,
+                        Message = e.Message,
+                        DiagnosticMessage = e.ToString(),
+                    };
+
+                    this.SendJsonRpcEventAsync(ScriptingErrorEvent.Type, eventParams);
+                }
+            }
+            finally
+            {
+                if (publishModel != null)
+                {
+                    publishModel.ScriptItemsCollected -= this.OnPublishModelScriptItemsCollected;
+                    publishModel.ScriptProgress -= this.OnPublishModelScriptProgress;
+                    publishModel.ScriptError -= this.OnPublishModelScriptError;
+                }
+            }
+        }
+
+        private void ValidateScriptDatabaseParams()
+        {
+            try
+            {
+                SqlConnectionStringBuilder builder = new SqlConnectionStringBuilder(this.Parameters.ConnectionString);
+            }
+            catch (Exception e)
+            {
+                throw new ArgumentException("Error parsing ConnectionString property", e);
+            }
+        }
+
+        private void OnPublishModelScriptError(object sender, ScriptEventArgs e)
+        {
+            ScriptingErrorParams eventParams = new ScriptingErrorParams
+            {
+                OperationId = this.OperationId,
+                Message = e.Error.Message,
+                DiagnosticMessage = e.Error.ToString(),
+            };
+
+            this.SendJsonRpcEventAsync(ScriptingErrorEvent.Type, eventParams);
+        }
+
+        private void OnPublishModelScriptItemsCollected(object sender, ScriptItemsArgs e)
+        {
+            if (this.cancellation.IsCancellationRequested)
+            {
+                throw new OperationCanceledException(this.cancellation.Token);
+            }
+
+            List<ScriptingObject> scriptingObjects = e.Urns.Select(urn => urn.ToScriptingObject()).ToList();
+            this.totalScriptedObjectCount = scriptingObjects.Count;
+
+            ScriptingPlanNotificationParams eventParams = new ScriptingPlanNotificationParams
+            {
+                OperationId = this.OperationId,
+                DatabaseObjects = scriptingObjects,
+                Count = scriptingObjects.Count,
+            };
+
+            this.SendJsonRpcEventAsync(ScriptingPlanNotificationEvent.Type, eventParams);
+        }
+
+        private void OnPublishModelScriptProgress(object sender, ScriptEventArgs e)
+        {
+            if (this.cancellation.IsCancellationRequested)
+            {
+                e.ContinueScripting = false;
+                throw new OperationCanceledException(this.cancellation.Token);
+            }
+
+            if (e.Completed)
+            {
+                this.scriptedObjectCount += 1;
+            }
+
+            // TODO: Handle the e.Error case.
```

Comment:
[https://github.com/Microsoft/sql-xplat-cli/issues/42](https://github.com/Microsoft/sql-xplat-cli/issues/42)

---
In reply to: [111653975](https://github.com/Microsoft/sqltoolsservice/pull/316#discussion_r111653975) [](ancestors = 111653975)